docs: autodoc_mock_imports for gssapi

ParthS007 opened the pull request:
closes #274

Adds mocking of  `gssapi` that cannot be installed on ReadTheDocs
builders.

This fixes the ReadTheDocs building problem.

diegodelemos commented on setup.py:
```diff
@@ -54,7 +54,6 @@
     "reana-db>=0.8.0a1,<0.9.0",
     "htcondor==8.9.7",
     "retrying>=1.3.3",
-    "paramiko[gssapi]>=2.6.0",
```
We cannot remove `paramiko` as a dependency otherwise the REANA-Job-Controller package will stop working [when it tries to use this library](https://github.com/reanahub/reana-job-controller/blob/5ac0d25ee128db8bbd51fe74f01ebfa11620a34e/reana_job_controller/utils.py#L83-L84).

Isn't it enough to include `autodoc_mock_imports` as you have done? Maybe we just need to find all the packages that make the build fail and add them to this list?

ParthS007 replied:
If I keep `paramiko[gssapi]` in `setup.py`, it gives an error because it tries to first install `gssapi`, but It works if I just keep `paramiko`.

diegodelemos replied:
We need to make it pass without modifying the content of `setup.py`. The current content of `setup.py` is necessary for REANA-Job-Controller to work correctly, therefore, we must find a solution elsewhere.

mvidalgarcia commented on setup.py:
```diff
@@ -54,7 +54,7 @@
     "htcondor==8.9.7",
     "retrying>=1.3.3",
-    "paramiko[gssapi]>=2.6.0",
+    "paramiko>=2.7.1",
```
We probably need to update `requirements.txt` via [`pip-compile`](https://github.com/jazzband/pip-tools) if we go for this change.

ParthS007 replied:
Yeah, if it's fine to just keep `paramiko`, I will make changes here as well. 👍

tiborsimko commented on setup.py:
```diff
@@ -32,6 +32,7 @@
         "sphinxcontrib-redoc>=1.5.1",
     ],
     "tests": tests_require,
+    "ssh": ["paramiko[gssapi]>=2.6.0"]
```
If `paramiko` *et al* needs to be moved out of installation requirements for ReadTheDocs to work, then we need to create a new file `requirements.in` where it shoulld be added, so that `pip-compile` would generate proper `requirements.txt` out of all our dependencies. You can test via `pip-compile --output-file=requirements.txt` and see what it includes

ParthS007 replied:
Adding it.

tiborsimko commented on setup.py:
```diff
@@ -54,7 +55,6 @@
     "reana-db>=0.8.0a1,<0.9.0",
     "htcondor==8.9.7",
     "retrying>=1.3.3",
-    "paramiko[gssapi]>=2.6.0",
```
Since `paramiko` became optional dependency, perhaps move import later into the function where it is really needed?

```
$ rg paramiko
reana_job_controller/utils.py
16:import paramiko
```

Chances are that nobody would `pip install r-j-controller` outside of REANA, so this should not matter too much... provided we always use `[all]`... but it could be good style.

(E.g. try a new virtual environment, install just like this `pip install .` and check `flake8`...)

tiborsimko commented:
@ParthS007 Can you please rebase all your RTFD build fixing branches against `maint-0.7` and edit PR to change the base branch? We can merge it there, so that we won'tt have to worry about failed PRs in case we shall be working on future 0.7.1, 0.7.2 bug fix upgrades...

ParthS007 commented:
@tiborsimko I have updated the base branch and rebased my current branch. Can you please check if it's fine and let me know if any other changes are required. 👍

tiborsimko commented on reana_job_controller/utils.py:
```diff
@@ -72,6 +71,8 @@ def initialize_krb5_token(workflow_uuid):
 class SSHClient:
     """SSH Client."""
 
+    import paramiko
```
If we import `paramiko` in this class, which is a good thing to do, we have to amend all paramiko-related commands to use `self.paramiko` as the package name, for example:

```diff
diff --git a/reana_job_controller/utils.py b/reana_job_controller/utils.py
index 9176495..418ba7a 100644
--- a/reana_job_controller/utils.py
+++ b/reana_job_controller/utils.py
@@ -81,8 +81,8 @@ class SSHClient:

     def establish_connection(self):
         """Establish the connection."""
-        self.ssh_client = paramiko.SSHClient()
-        self.ssh_client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+        self.ssh_client = self.paramiko.SSHClient()
+        self.ssh_client.set_missing_host_key_policy(self.paramiko.AutoAddPolicy())
         self.ssh_client.connect(hostname=self.hostname, port=self.port, gss_auth=True)

     def exec_command(self, command):

```

ParthS007 replied:
Done 👍

tiborsimko commented on requirements.in:
```diff
@@ -0,0 +1,7 @@
+# This file is part of REANA.
+# Copyright (C) 2017, 2018 CERN.
```
2020

(we did not have this file in 2017, 2018)

ParthS007 replied:
Done 👍

tiborsimko replied:
You edited the wrong file, the 2020 should be in the new `requirements.in` file, not in the old `MANIFEST.in` ...
